src/Features/VisualBasic/Portable/RemoveAsyncModifier/VisualBasicRemoveAsyncModifierCodeFixProvider.vb

```diff
@@ -0,0 +1,111 @@
+
+        Private Shared ReadOnly s_diagnosticIds As ImmutableArray(Of String) = ImmutableArray.Create(BC42356)
+
+        <ImportingConstructor>
+        <SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification:="Used in test code: https://github.com/dotnet/roslyn/issues/42814")>
+        Public Sub New()
+        End Sub
+
+        Public Overrides ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String)
+            Get
+                Return s_diagnosticIds
+            End Get
+        End Property
+
+        Protected Overrides Function IsAsyncSupportingFunctionSyntax(node As SyntaxNode) As Boolean
+            Return node.IsAsyncSupportedFunctionSyntax()
+        End Function
+
+        Protected Overrides Function ShouldOfferFix(methodSymbol As IMethodSymbol, knownTypes As KnownTypes) As Boolean
+            ' We only apply to functions because async subs will be offered the Make Method Synchronous fixer which would do the same job
+            Return Not methodSymbol.ReturnsVoid
+        End Function
+
+        Protected Overrides Function RemoveAsyncModifier(methodSymbolOpt As IMethodSymbol, node As SyntaxNode, knownTypes As KnownTypes) As SyntaxNode
+            Dim methodBlock = TryCast(node, MethodBlockSyntax)
+            If methodBlock IsNot Nothing Then
+                Dim subOrFunctionStatement = methodBlock.SubOrFunctionStatement
+                Dim newSubOrFunctionStatement = RemoveAsyncModifierHelpers.RemoveAsyncKeyword(subOrFunctionStatement)
+                Return methodBlock.WithSubOrFunctionStatement(newSubOrFunctionStatement)
+            End If
+
+            Dim multiLineLambda = TryCast(node, MultiLineLambdaExpressionSyntax)
+            If multiLineLambda IsNot Nothing Then
+                Return RemoveAsyncModifierHelpers.FixMultiLineLambdaExpression(multiLineLambda)
+            End If
+
+            Dim singleLineLambda = TryCast(node, SingleLineLambdaExpressionSyntax)
+            If singleLineLambda IsNot Nothing Then
+                Return RemoveAsyncModifierHelpers.FixSingleLineLambdaExpression(singleLineLambda)
+            End If
+
+            Return Nothing
+        End Function
+
+        Protected Overrides Function AnalyzeControlFlow(semanticModel As SemanticModel, node As SyntaxNode) As ControlFlowAnalysis
+
+            Dim methodBlock = TryCast(node, MethodBlockSyntax)
+            If methodBlock IsNot Nothing Then
+                Dim statements = methodBlock.Statements
+                Return semanticModel.AnalyzeControlFlow(statements.First(), statements.Last())
+            End If
+
+            Dim multiLineLambda = TryCast(node, MultiLineLambdaExpressionSyntax)
+            If multiLineLambda IsNot Nothing Then
+                Dim statements = multiLineLambda.Statements
+                Return semanticModel.AnalyzeControlFlow(statements.First(), statements.Last())
+            End If
+
+            Return Nothing
```

Comment:
is the caller resilient to getting null back?

Reply:
Yes